codegen/src/main/java/io/reactiverse/es4x/codegen/generator/IndexDTS.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018 Paulo Lopes.
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *  The Eclipse Public License is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  The Apache License v2.0 is available at
+ *  http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+package io.reactiverse.es4x.codegen.generator;
+
+import io.vertx.codegen.*;
+import io.vertx.codegen.type.ApiTypeInfo;
+import io.vertx.codegen.type.ClassTypeInfo;
+import io.vertx.codegen.type.EnumTypeInfo;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.*;
+
+import static io.reactiverse.es4x.codegen.generator.Util.*;
+
+public class IndexDTS extends Generator<ClassModel> {
+
+  public IndexDTS() {
+    incremental = true;
+
+    kinds = new HashSet<>();
+    kinds.add("class");
+
+    name = "es4x-generator";
+  }
+
+  @Override
+  public String filename(ClassModel model) {
+    return "npm/index.d.ts";
+  }
+
+  @Override
+  public String render(ClassModel model, int index, int size, Map<String, Object> session) {
+
+    StringWriter sw = new StringWriter();
+    PrintWriter writer = new PrintWriter(sw);
+
+    ClassTypeInfo type = model.getType();
+
+    if (index == 0) {
+      Util.generateLicense(writer);
+
+      if (type.getModuleName().equals("vertx")) {
+        writer.print("export interface Handler<T> {\n");
```

Comment:
Granted, it looks like its too late :-) but FWIW if you end up refactoring this again, you might look at:

https://github.com/outfoxx/typescriptpoet/blob/develop/src/test/java/io/outfoxx/typescriptpoet/test/ClassSpecTests.kt#L33

It's an adaption of Square's javapoet library/pattern to generate TS source files (from the JVM, which is somewhat ironic, but I assume b/c outfoxx has a JVM-based codegen tool chain).

I haven't actually used it but in theory you get the benefits of the poet pattern, e.g. auto-management of imports, ~generally great formatting/indentation handled for you/etc.